Migrate to code blocks to import from .re source files

feihong opened the pull request:
After copying all the book's markdown files to vitepress site, I migrate almost all code blocks inside the markdown files to import from file by region.

feihong commented on docs/celsius-converter-exception/Snippets.re:
```diff
@@ -0,0 +1,234 @@
+  let convert = x => x;
+  // #region pipes
+  <div>
+    <input /* ... */ />
```
Fortunately `/* */` comments inside a component don't get shuffled around in JSX expressions

feihong commented on docs/celsius-converter-exception/index.md:
````diff
@@ -4,26 +4,7 @@ Our second widget will be one that takes a temperature value in Celsius and
-  </div>;
-};
-```
+<<< Snippets.re#celsius-converter-v1
````
I guess a small disadvantage is that when you edit these docs, you have to keep the dev server open in a tab so you can see what it renders

jchavarri replied:
True, but it doesn't looks like a big deal, as looking at the document rendered by the server is generally the "final proof" to make sure everything is ok. Similar thing happens with melange.re and mkdocs.

feihong commented on docs/celsius-converter-option/Snippets.re:
```diff
@@ -0,0 +1,107 @@
+  | Some(fahrenheit) =>
+    (fahrenheit |> convert |> Js.Float.toFixedWithPrecision(~digits=2))
+    ++ {js|°F|js}
+  };
```
I'm not completely satisfied with the snippets in this file because they end with `;`, which, if pasted directly into the user's code, will cause a syntax error. One workaround is to include the entire JSX expression, but that makes for overly-long code blocks.

I guess the way to handle it is to make it so files in docs directory can't be auto-formatted using `dune fmt` and then include a comment at the top of this file warning against hand-editing. Wdyt?

feihong replied:
Thought about this and my current plan is to split my snippets into two separate files: Snippets.re (for ones that can be auto-formatted), and NoFmtSnippets.re (for those that shouldn't be autoformatted). @jchavarri Wdyt?

jchavarri replied:
> if pasted directly into the user's code, will cause a syntax error.

Sorry, I am confused about this, could you elaborate? The way I see it, if they copy just the code inside the region, it's valid Reason snippet. If they copy the whole `let _ = {` (until the closing bracket) it's also valid syntax.

How will it cause a syntax error?

> Thought about this and my current plan is to split my snippets into two separate files: Snippets.re (for ones that can be auto-formatted), and NoFmtSnippets.re (for those that shouldn't be autoformatted).

This sounds quite convoluted to me.

feihong replied:
These snippets won't compile if they are pasted in because they're part of a larger expression which looks something like this:
```reason 
    {(
       String.trim(celsius) == ""
         ? {js|?°F|js}
         : (
           switch (celsius |> float_of_string_opt |> Option.map(convert)) {
           | None => "error"
           | Some(fahrenheit) when fahrenheit < (-128.6) => {js|Unreasonably cold🥶|js}
           | Some(fahrenheit) when fahrenheit > 212.0 => {js|Unreasonably hot🥵|js}
           | Some(fahrenheit) =>
             Js.Float.toFixedWithPrecision(fahrenheit, ~digits=2)
             ++ {js|°F|js}
           }
         )
     )
     |> React.string}
```
So the trailing `;` will break the expression when you substitute the switch expression. The problem is that refmt will always reorder comments placed directly after a switch expression, making it impossible to get the boundaries of the region right unless autoformat isn't used on the file.

feihong replied:
After a bit of experimentation, I found this hacky workaround to prevent refmt from moving my comments around:
```reason
  // #region float-of-string-opt
  (
    switch (celsius |> float_of_string_opt |> Option.map(convert)) {
    | None => "error"
    | Some(fahrenheit) when fahrenheit < (-128.6) => {js|Unreasonably cold🥶|js}
    | Some(fahrenheit) when fahrenheit > 212.0 => {js|Unreasonably hot🥵|js}
    | Some(fahrenheit) =>
      Js.Float.toFixedWithPrecision(fahrenheit, ~digits=2) ++ {js|°F|js}
    }
  )
  // #endregion float-of-string-opt
  |> ignore;
```
There's a slight downside of extra parentheses being included in the snippet, but copying and pasting those parentheses shouldn't cause compiler errors.

jchavarri replied:
The approach with parenthesis sounds good to me. Another alternative I was thinking about is to just inline these problematic snippets in the markdown... If there are not too many of them, then it should be ok.

feihong replied:
> inline these problematic snippets in the markdown

It's likely I'll end up doing that in the future, but this refmt issue applies to all the snippets in this particular file :see_no_evil:

feihong committed:
Migrate code snippets to import from Snippet.re files

feihong commented on docs/counter/index.md:
````diff
@@ -103,67 +75,24 @@ same as the name of the file.
-let _ =
--->
 ```reason
 {counter |> Int.to_string |> React.string}
````
I decided to leave these tiny blocks alone and not import them from file

jchavarri commented on docs/celsius-converter-exception/dune:
```diff
@@ -0,0 +1,6 @@
+(melange.emit
```
An alternative "design" for this could be to have just `library` stanzas on each chapter folder, with the `(modes melange)`, and then in `docs/dune` have a single `melange.emit` that includes all the libraries.

The main upside is that things like `module_systems` or `target` are only defined once.

feihong replied:
I think I'll keep the current layout for now. Some chapters will explicitly be about creating libraries and not apps, so I want to keep the distinction clear.

jchavarri commented on docs/celsius-converter-exception/index.md:
```diff
@@ -38,19 +19,13 @@ But beware, because the compiler knows nothing about the types of those fields.
 For example, we could write `ReactEvent.Form.target(evt)##value + 1`, treating
 it as if it were an integer, and the compiler wouldn't complain.
 
+## Wrap functions that return Js.t
```
```suggestion
## Wrap functions that return `Js.t`
```
or maybe "a `Js.t` value"?

feihong replied:
When possible, I try to keep the subsection headers short so that they aren't truncated in the sidebar

jchavarri commented on docs/celsius-converter-exception/index.md:
```diff
@@ -61,16 +36,12 @@ body of the callback function is surrounded by braces (`{}`). OCaml functions
 need to be enclosed by braces.
 
-Let's change the render logic to update the Fahrenheit display:
+## Apply pipe last (|>)
```
```suggestion
## Apply pipe last (`|>`)
```

jchavarri commented on docs/celsius-converter-exception/index.md:
````diff
@@ -81,89 +52,41 @@ We should probably put °F after the Fahrenheit value so that it's clear to the
-```
+## String concatenation (++)
+
+<<< Snippets.re#string-concat{4-6}
````
This kind of reference to lines sounds like something that could easily break in the future. I guess the workflow when changing snippets would be:
- Change snippet
- Grep for usages of the snippet region and update them
?

feihong replied:
Yeah, and I have an unofficial rule to never use a snippet twice in one place. It would probably be redundant anyway, but it could easily break the line highlighting if I'm not careful.

jchavarri commented on src/melange-playground/Makefile:
```diff
@@ -0,0 +1,2 @@
+serve:
+	chapter=melange-playground make -C ../.. serve
```
~Are the Makefiles in these folders useful anymore? I thought one would just call `npm run docs:dev` (or `bun run docs:dev`) and have access to all chapters at once.~

Edit: oh, I see, these are to run the actual code of the book. And the `chapter` is read from top level command.

jchavarri commented on docs/counter/Snippets.re:
```diff
@@ -0,0 +1,65 @@
+// #region use-app-component
+let node = ReactDOM.querySelector("#root");
+switch (node) {
+| Some(root) => ReactDOM.render(<App />, root)
```
Do you know why `root` is underlined?

<img width="713" alt="image" src="https://github.com/melange-re/melange-for-react-devs/assets/220424/0cf47f50-df55-4521-b2da-9a560f67ed14">

feihong replied:
No idea. Do you think I should try to fix it? It looks like anything that is pattern-matched is underlined.

jchavarri replied:
I don't know. Did you get the grammar from the vscode-ocaml extension? I can't see that behavior in vscode itself.

feihong replied:
> Did you get the grammar from the vscode-ocaml extension?

https://github.com/melange-re/melange-for-react-devs/blob/1a4b87d0f374dcf21525b8973d64143fae0bcf72/docs/.vitepress/config.js#L5-L6

For reference, it looks like anything that appears as the lighter red color in VS Code (in default light theme) will be underlined in vitepress syntax highlighting.

feihong replied:
I'll try playing around with shiki config to see if there's a theme that doesn't underline these parts. Do you think it's a blocker?

feihong replied:
Looks like the default themes applied are github-light and github-dark, and those seem to be the only themes that apply underlining. None of the other built-in themes I checked seem to have it. So we can easily switch to one of those.

jchavarri replied:
Sounds good. The underlining is a bit distracting, but it might be just me.

jchavarri commented on docs/celsius-converter-option/index.md:
````diff
@@ -258,76 +160,22 @@ string to float. Hint: Use `Js.Float.isNaN`.
-)
-|> React.string}
-```
+<<< Snippets.re#string-trim{1-2}
````
This highlights the second line too, with the `ºF` part, not sure if intended

<img width="751" alt="image" src="https://github.com/melange-re/melange-for-react-devs/assets/220424/ecf7f9f9-c80a-4966-9489-57833860f0e4">

feihong replied:
This was intentional, to emphasize that both lines are new additions to the original expression

jchavarri commented on docs/celsius-converter-exception/index.md:
````diff
@@ -81,89 +52,41 @@ We should probably put °F after the Fahrenheit value so that it's clear to the
-```reason
-{(celsius |> float_of_string |> convert |> string_of_float) ++ {js| °F|js} |> React.string}
-```
+## String concatenation (++)
````
```suggestion
## String concatenation (`++`)
```

jchavarri reviewed (commented):
It looks great! I have a couple more things, besides what I mentioned on inline comments:

- Is `bun` necessary? the top level `package.json` is supposed to be used with npm (there is a `package-lock.json`. I think we should use a single package manager, and Bun is still very alpha (imo)
- Will some kind of search be available in the docs?
- Can there be "Edit this page on GitHub" buttons in Vitepress?
- There is some comment `> Docendo disco, scribendo cogito` in the todo.md page. Is that on purpose?
- Should there be some explanation at the top level readme to explain the separation between book and sources?

feihong committed:
Add search bar

feihong committed:
Use npm instead of bun

feihong committed:
Add 'Edit this page' link

feihong committed:
Put backticks around code elements in subtitles

feihong committed:
Add a note about the book files in top README

feihong commented:
> Is bun necessary? the top level package.json is supposed to be used with npm (there is a package-lock.json. I think we should use a single package manager, and Bun is still very alpha (imo)

Switched to npm in https://github.com/melange-re/melange-for-react-devs/pull/1/commits/698f1475540ab3ba14203053a192207bc395022a 

> Will some kind of search be available in the docs?

Enabled the search bar in https://github.com/melange-re/melange-for-react-devs/pull/1/commits/024668860ec81668d8d7c612dd977b1c8ce8d154

> Can there be "Edit this page on GitHub" buttons in Vitepress?

Added "edit this page" link at the bottom of every page in https://github.com/melange-re/melange-for-react-devs/pull/1/commits/58b8d43a651d70a3b3e44f30e4c0378264d74a91

> There is some comment > Docendo disco, scribendo cogito in the todo.md page. Is that on purpose?

Yes, it means "I learn by teaching, I think by writing" in Latin.

> Should there be some explanation at the top level readme to explain the separation between book and sources?

Added note that books files are in docs directory in https://github.com/melange-re/melange-for-react-devs/pull/1/commits/37b2a092bde45f50ca3336f53282080056f5e1c7

jchavarri reviewed (approved):
Looks great!

feihong committed:
Fix snippets so they won't cause syntax errors if they're copied and pasted into reader's code
